### PR TITLE
V2 Modernize boilerplate

### PR DESCRIPTION
**Pull request by jayqi**
Here is a stab at modernizing the boilerplate.

- Adds a `config` module which:
    - Runs `load_dotenv`
    - Sets `PROJ_ROOT`
    - Creates a bunch of other Paths for project directories
    - Configures loguru to work with tqdm if tqdm is installed
- Uses typer instead of click for the CLIs.
- Renames some of the modules.
- Adds some equivalent CLI boilerplate as the data one for features, model training, model predicting, and viz. Previously, these were blank files. 

Open to opinions here. I also considered just getting rid of all of the CLI example modules.

### CODE REVIEW

File: {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/models/predict.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from pathlib import Path
+
+from loguru import logger
+from tqdm import tqdm
+import typer
```
Comment (pjbull) [inline {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/models/predict.py:5]: Do I recall correctly that typer is nicer to use if `rich` is also in the env (but it is an optional depdency)? Should we include that as well?

**Review by pjbull — COMMENTED** [review]
Awesome! Big improvement here.

A little on the fence with a slight preference for `dataset.py`, `features.py` and `plot.py` over the ones that start `some_`. Removes the rename step and people can customize as desired if they want but the default doesn't require it.

Can we have some comment fence around the filler code that should be replaced? Something like:

```python
# REPLACE THE BELOW WITH YOUR CODE ----------------------------------
logger.info("Processing some dataset...")
for i in tqdm(range(10), total=10):
    if i == 5:
        logger.info("Something happened for iteration 5.")
logger.success("Processing some dataset complete.")
# -------------------------------------------------------------------
```

Some nice-to-haves that can be issues are:
 - Docs about how to change the boilerplate
 - Tests that the typer commands execute
 - Boilerplate tests.py file that runs with pytest

File: {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/config.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        dist = importlib.metadata.distribution("{{ cookiecutter.module_name }}")
+        direct_url_data = dist.read_text("direct_url.json")
+        if direct_url_data is None:
+            return False
+        return json.loads(direct_url_data).get("dir_info", {}).get("editable", False)
+    except importlib.metadata.PackageNotFoundError:
+        return False
+
+
+IS_EDITABLE = _is_editable()
```
[the remark this answers is not in the document]
Comment (jayqi) [reply {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/config.py:27]: @pjbull I disagree here, and this is from personal experience. In a volunteer data project I contributed to recently, someone had originally documented that the way to install the source was to do `pip install .`. I think this is an easy situation for someone to end up in, and would totally break the paths.

[the remark this answers is not in the document]
Comment (pjbull) [reply {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/config.py:27]: I'm still not 100% convinced for three reasons:
 - It's a lot of esoteric code to put in the "user space" (and also additional code to support/maintain)
 - We include the editable install in the [dependencies](https://github.com/drivendata/cookiecutter-data-science/blob/v2/ccds/hook_utils/dependencies.py) for whatever manager you choose.
 - The errors would be loud (and documentable, thus googleable) if something goes wrong, and the fix is easy.

File: {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/models/predict.py (hunk 1 above)
Comment (jayqi) [reply {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/models/predict.py:5]: As of 0.12.1 (released last week), typer now includes the optional dependencies by default. You'd instead install `typer-slim` to get the minimal version. 

https://typer.tiangolo.com/release-notes/#0121

**Comment by jayqi** [comment]
> A little on the fence with a slight preference for dataset.py, features.py and plot.py over the ones that start some_. Removes the rename step and people can customize as desired if they want but the default doesn't require it.

The reason I like the `some_` prefix is that it's a sign post that the name is not special or expected to be final and that people _should_ pick a real name for it. 

I don't really like having a default generic name. If someone were really going to just have a single `dataset` module, then having `data.dataset`, `features.features`, etc. is weird and we should just get rid of the directory layers.

File: {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/data/some_dataset.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+from {{ cookiecutter.module_name }}.config import RAW_DATA_DIR, PROCESSED_DATA_DIR
+
+app = typer.Typer()
+
+
+@app.command()
+def main(
+    input_path: Path = RAW_DATA_DIR / "some_dataset.csv",
+    output_path: Path = PROCESSED_DATA_DIR / "some_dataset.csv",
```
[the remark this answers is not in the document]
Comment (jayqi) [reply {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/data/some_dataset.py:15]: I was a bit on the fence about this one. My reason was that maybe the processed data is useful on its own. Maybe the project isn't doing any modeling and it's just analysis.

I could go either way on this though.

**Comment by jayqi** [comment]
Well okay. Since we've got some real stuff in the boilerplate now with imports like

```
from {{ cookiecutter.module_name }}.config import FIGURES_DIR, PROCESSED_DATA_DIR
```

this breaks linting and formatting. 

Options:

1. Use relative imports to avoid the template syntax. I don't like relative imports (and PEP 8 recommends absolute imports too) so I don't like this. 
2. Move linting into tests. This would work for linting, but wouldn't help with formatting. **This is my leaning.**
3. Get rid of the boilerplate so we don't have to deal with this.

File: {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/data/some_dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pjbull) [reply {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/data/some_dataset.py:15]: Yeah, I was also on the fence. I think this is fine as is.

**Comment by pjbull** [comment]
Agreed about #2 for linting.

File: tests/test_creation.py [hunk 4]
```diff
@@ -183,3 +183,29 @@ def verify_makefile_commands(root, config):
     assert "clean                    Delete all compiled Python files" in stdout_output
 
     assert result_returncode == 0
+
+def lint(root):
```
Comment (pjbull) [inline tests/test_creation.py:189]: Where does this get called?

**Review by pjbull — COMMENTED** [review]
~Looks great. Only thing I'm wondering is where the new `lint` function gets called.~

File: tests/test_creation.py (hunk 4 above)
Comment (pjbull) [reply tests/test_creation.py:189]: Oops, reviewed too soon

**Comment by jayqi** [comment]
@pjbull Should be good to go! I opened these two issues to capture some of the items we're punting on:

- #356 
- #357

**Pull request merged**